packages/tailwindcss/src/theme-fn.ts

```diff
@@ -0,0 +1,152 @@
+import { deepMerge } from './compat/config/deep-merge'
+import type { UserConfig } from './compat/config/types'
+import type { DesignSystem } from './design-system'
+import type { Theme } from './theme'
+import { DefaultMap } from './utils/default-map'
+import { toKeyPath } from './utils/to-key-path'
+
+export function createThemeFn(
+  designSystem: DesignSystem,
+  configTheme: () => UserConfig['theme'],
+  resolveValue: (value: any) => any,
+) {
+  return function theme(path: string) {
+    let keypath = toKeyPath(path)
+    let cssValue = readFromCss(designSystem.theme, keypath)
+
+    // If we get a single non-null value back from the CSS theme, return it
+    // Otherwise, we need to merge the CSS theme with the one resolved from
+    // all registered plugins.
+    if (cssValue === null || typeof cssValue !== 'object') {
+      return cssValue
+    }
+
+    let result: Record<string, any> = configTheme() ?? {}
+
+    let configValue = resolveValue(get(result, keypath) ?? {})
+
+    if (typeof configValue === 'object') {
+      deepMerge(configValue, [cssValue], (_, b) => b)
+      set(result, keypath, configValue)
+
+      return get(result, keypath)
```

Comment:
I am concerned that mutating the `configValue` from inside a function to _read_ from the config is going to create some unforeseen and hard-to-reason-about side effects (I couldn't come up with an issue right now but it does mean that calling `theme()` multiple times will update the plugin provided config over and over which seems really unexpected?). We don't need to do this, `return configValue` after the `deepMerge` also makes all tests pass.

I also saw that `deepMerge` mutates the input object though. Right now this is somewhat hidden because tests often use the function syntax for theme values so that every invocation would create a new object but for literal objects this could be an issue. Maybe we can instead create a new object here regardless and do something like this?

```suggestion
      return deepMerge({}, [configValue, cssValue], (_, b) => b)
```

Reply:
Done